### PR TITLE
ssh/tailssh,util: extract new osuser package from ssh code

### PR DESCRIPTION
**Pull request by awly**
This package is a wrapper for os/user that handles non-cgo builds, gokrazy and user shells.

Updates tailscale/corp#15405

### CODE REVIEW

File: util/osuser/user.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	"tailscale.com/version/distro"
+)
+
+// LookupId is like os/user.LookupId but handles a few edge cases like gokrazy
+// and non-cgo lookups, and returns the user shell.
+func LookupId(uid string) (u *user.User, shell string, err error) {
```
Comment (bradfitz) [inline util/osuser/user.go:25]: LookupId is a historical spelling mistake we can't fix now, but we wish it had really been LookupID per https://github.com/golang/go/wiki/CodeReviewComments#initialisms

There are arguments for either way here (match os/user vs use modern style). Your call.

File: util/osuser/user.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	"runtime"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	"tailscale.com/version/distro"
+)
+
+// LookupId is like os/user.LookupId but handles a few edge cases like gokrazy
+// and non-cgo lookups, and returns the user shell.
```
Comment (bradfitz) [inline util/osuser/user.go:24]: Clarify that this returned shell may be empty and that it's best effort.

File: util/osuser/user.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		Username: f[0],
+		Uid:      f[2],
+		Gid:      f[3],
+		Name:     f[4],
+		HomeDir:  f[5],
+	}, f[6], nil
+}
+
+func fallbackShell() string {
+	if s := os.Getenv("SHELL"); s != "" {
```
Comment (bradfitz) [inline util/osuser/user.go:111]: This doesn't seem right. If I'm looking up a different user on the box, my current SHELL shouldn't influence the info returned about some other user.

This is only applicable if the User we're returning matches os/user.Current() (or is non-Windows and has UID of os.Getuid)

File: util/osuser/user.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	for _, r := range usernameOrUID {
+		if r < ' ' || r == 0x7f || r == utf8.RuneError { // TODO(bradfitz): more?
+			return false
+		}
+	}
+	return true
+}
+
+func userLookupGetent(usernameOrUID string, lookupStd func(string) (*user.User, error)) (*user.User, string, error) {
```
Comment (bradfitz) [inline util/osuser/user.go:82]: all these "usernameOrUID" sound like these funcs are overloaded and accept either one so I was looking at first where the magic was that heuristically decides whether it's a UID vs a username (and then worrying about security problems with said heuristic) and then I realized what's going on: that user.LookupId also takes a string and you wanted the same wrapper.

Maybe `lookupFuncArg` in general?

But in this case, isn't it always a username and not a UID?

File: util/osuser/user.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+		return u, shell, nil
+	}
+
+	// On Linux, default to using "getent" to look up users so that
+	// even with static tailscaled binaries without cgo (as we distribute),
+	// we can still look up PAM/NSS users which the standard library's
+	// os/user without cgo won't get (because of no libc hooks).
+	// But if "getent" fails, userLookupGetent falls back to the standard
+	// library anyway.
+	return userLookupGetent(usernameOrUID, lookupStd)
```
Comment (bradfitz) [inline util/osuser/user.go:63]: on Linux, can we default to trying os/user first and only shell out to a child process on failure?

I guess the difference is we don't know whether the caller wants the shell info or not. Some callers do; some don't. If they do (SSH), then we have to do getent first. If they don't, it'd be better to not make child processes unnecessarily on 99% of the machines out there.

Maybe we need a separate top-level func(s) like LookupUserAndShell to declare you want the shell and only Tailscale SSH callers can use that?

File: util/osuser/user.go (hunk 4 above)
Comment (awly) [reply util/osuser/user.go]: I'll need lookup by UID here: https://github.com/tailscale/tailscale/pull/10163/files#r1385655690
As far as ambiguity - `getent` handles both username and UIDs. Can you elaborate on `lookupFuncArg`?

File: util/osuser/user.go (hunk 3 above)
Comment (awly) [reply util/osuser/user.go]: Yeah, this was taken from ssh code and is probably not right here.

File: util/osuser/user.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	"runtime"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	"tailscale.com/version/distro"
+)
+
+// LookupUIDAndShell is like os/user.LookupId but handles a few edge cases like
+// gokrazy and non-cgo lookups, and returns the user shell. User shell lookup
```
Comment (bradfitz) [inline util/osuser/user.go:23]: The user shell lookup ....

(here and L30)

File: util/osuser/user.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	"os/user"
+	"runtime"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	"tailscale.com/version/distro"
+)
+
+// LookupUIDAndShell is like os/user.LookupId but handles a few edge cases like
```
Comment (bradfitz) [inline util/osuser/user.go:22]: "And" makes this name a bit weird now. "LookupUID" really means "Lookup by UID" and the "by" was implicit. That worked before. But once you say "LookupUIDAndShell", you're not looking up _by_ the UID and Shell, and not looking up _by_ the shell, so it now reads like you're looking up the UID and looking up the shell, but we're not looking up the UID.

Maybe "osuser.LookupWithShellByUID" and "osuser.LookupWithShellByUsername"? Kinda verbose, but more clear?

File: util/osuser/user.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// LookupUsernameAndShell is like os/user.Lookup but handles a few edge cases
+// like gokrazy and non-cgo lookups, and returns the user shell. User shell
+// lookup is best-effort and may be empty.
+func LookupUsernameAndShell(username string) (u *user.User, shell string, err error) {
+	return lookup(username, user.Lookup, true)
+}
+
+// LookupUIDAndShell is like os/user.LookupId but handles a few edge cases like
```
Comment (bradfitz) [inline util/osuser/user.go:36]: LookupUID

File: util/osuser/user.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+	// If shell is not required, try os/user.Lookup* first and only use getent
+	// if that fails. This avoids spawning a child process when os/user lookup
+	// succeeds.
+	if u, err := std(usernameOrUID); err == nil {
+		return u, "", nil
+	}
+	return userLookupGetent(usernameOrUID, std)
+}
+
+func validUsername(usernameOrUID string) bool {
```
Comment (bradfitz) [inline util/osuser/user.go:91]: weird func name given the argument name.

It's not clear what this is trying to do: validate usernames or validate userIDs? The caller is using it for both, but originally it was only for usernames.

It happens to mostly work for userIDs too, but it's not super strict. At least it prevents passing totally shady stuff to getent on UID lookups.

Maybe rename to "checkGetentInput"?

File: util/osuser/user.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+func LookupUsername(username string) (*user.User, error) {
+	u, _, err := lookup(username, user.Lookup, false)
+	return u, err
+}
+
+// lookupStd is either user.Lookup or user.LookupId.
+type lookupStd func(string) (*user.User, error)
+
+func lookup(usernameOrUID string, std lookupStd, wantShell bool) (*user.User, string, error) {
+	if runtime.GOOS != "linux" {
```
Comment (bradfitz) [inline util/osuser/user.go:54]: we want to use getent on any other platforms? FreeBSD at least has this?

Maybe a TODO?

File: util/osuser/user.go (hunk 4 above)
Comment (bradfitz) [reply util/osuser/user.go]: > As far as ambiguity - getent handles both username and UIDs. 

Ah, cool. I see that in the manpage now. And we use it for UID lookups now too. Got it.

> Can you elaborate on lookupFuncArg

No need now.

**Review by bradfitz — APPROVED** [review]

File: util/osuser/user.go (hunk 10 above)
Comment (awly) [reply util/osuser/user.go:55]: Makes sense, left a TODO for now

**Pull request merged**